org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/OrganizeImportsCommand.java

```diff
@@ -77,22 +75,30 @@ public static Object organizeImports(List<Object> arguments) throws CoreExceptio
 			}
 			// Project root URI:
 			if (targetProj != null && selectProjectRoot) {
-				return command.organizeImportsInProject(targetProj);
-			}
-
-			URI uri = URI.create(fileUri);
-			File file = new File(uri.getPath());
-			boolean exists = file.exists();
-			if (!exists) {
-				return new WorkspaceEdit();
-			}
-			if (file.isDirectory()) {
-				return command.organizeImportsInDirectory(fileUri, targetProj);
+				edit = command.organizeImportsInProject(targetProj);
 			} else {
-				return command.organizeImportsInFile(fileUri);
+				URI uri = URI.create(fileUri);
+				File file = new File(uri.getPath());
+				if (file.exists()) {
+					if (file.isDirectory()) {
+						edit = command.organizeImportsInDirectory(fileUri, targetProj);
+					} else {
+						edit = command.organizeImportsInFile(fileUri);
+					}
+				}
 			}
 		}
-		return new WorkspaceEdit();
+		final boolean applyNow = JavaLanguageServerPlugin.getPreferencesManager().getClientPreferences().isWorkspaceApplyEditSupported();
+		if (applyNow) {
+			JavaLanguageServerPlugin.getInstance().getClientConnection().applyWorkspaceEdit(edit);
+			// return an empty object to avoid errors on client
+			return new Object();

```

Comment:
This just looks fishy to me. Sometimes jdt.ls handles the change and sometimes the client?

Reply:
It is not sometimes it depends on the capability of the client. If workspace/applyEdit is not supported by the client then we just send the workspaceEdit and expect the command to handle it. This change decreases our requirements on the client integration for these commands to work.